plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginService.java

```diff
@@ -28,17 +44,97 @@
  */
 public class GoogleLoginService {
 
-  // TODO(chanseok): this will be needed later.
-  public static final List<String> OAUTH_SCOPES = Collections.unmodifiableList(Arrays.asList(
-      "https://www.googleapis.com/auth/userinfo#email"
+  private static final String STASH_OAUTH_CRED_KEY = "OAUTH_CRED";
+  private static final List<String> OAUTH_SCOPES = Collections.unmodifiableList(Arrays.asList(
+      "email", //$NON-NLS-1$
+      "https://www.googleapis.com/auth/cloud-platform" //$NON-NLS-1$
   ));
 
   /**
-   * Returns the credential of the active user. If there is no active user, returns {@code null}.
+   * Returns the credential of an active user (among multiple logged-in users). A login screen
+   * may be presented, e.g., if no user is logged in or login is required due to an expired
+   * credential.
+   *
+   * @param shellProvider provides a shell for the login screen if login is necessary
+   * @throws IOException can be thrown by the underlying Login API request
+   */
+  // TODO(chanseok); synchronize properly
```

Comment:
This has always bothered me. It seems like virtually all Eclipse (and Intellij too) code doesn't care about synchronization. For example, looking at their code, I think none of `getWorkbench()`, `getService()`, `getTransientData()`,`get()`, and `put()` is thread-safe. I think it's mostly acceptable given that Eclipse is not a mission-critical system, the chances of contention is very low, and the consequences from unsafe concurrent operations are tolerable.

In this regard, I think I'll drop the comment that `GoogleLoginService` is thread-safe and follow the general practice that we don't synchronize most of the time. Although `put()` is not atomic and thus concurrent puts may corrupt the transient map in theory, I think in practice the possibility is zero. Even if the map has been corrupted, users can easily recover from malfunction, e.g., by restarting Eclipse.